### PR TITLE
replace sparkline library with modern equivalent

### LINKED ISSUE #12065: PHP7.2 no longer supports 'object' as a class name in Sparkline
In `piwik/libs/sparkline/lib/Object.php`, the class 'Object' should be renamed.

### LINKED ISSUE #12271: When there is no data to show in the sparkline, it shows nothing instead of showing a flat sparkline
When there is no data to show in the sparkline, it shows nothing instead of showing a flat sparkline:

## No data

empty heatmaps

## Another example

another example

This may be fixed with https://github.com/piwik/piwik/pull/12066 ?

### PR DESCRIPTION
**Pull request by Findus23**
Fixes #12065
Fixes #12271 
Refs #8410

The original sparkline library wasn't updated for 12 years. The original author has written a new version, but it is intended to be used as a standalone file: https://github.com/jamiebicknell/Sparkline

But there is another modern php library which creates similar sparklines: 
https://github.com/davaxi/Sparkline

I also tried to improve the color scheme but I am open for improvements.

Unfortunately it doesn't support adding dots, so there are no minimum/maximum points anymore (But I am not sure how useful they are)

This definitely needs some testing before merging, as I don't know all variants of sparklines in piwik.

### CODE REVIEW

**Comment by Findus23** [comment]
Oddly the only failing tests have nothing to do with this change.
are there no ui tests of the all websites overview?

**Comment by Findus23** [comment]
It seems like the sparklines aren't visible on the UI tests.

(tests/UI/expected-screenshots/MultiSitesTest_all_websites.png)

**Comment by Findus23** [comment]
Seems like now random other things aren't working:
http://builds-artifacts.piwik.org/piwik/piwik/sparklines/24357/Live_visitor_profile_actions_hidden.png

**Comment by Findus23** [comment]
And I guess there was a reason for `\Piwik\Visualization\Sparkline::$enableSparklineImages = false;`
as the test ran endlessly:
https://travis-ci.org/piwik/piwik/builds/276259820

**Comment by mattab** [comment]
Hi @Findus23 

Feedback

* the reason the sparklines are disabled in UI tests is because GD library generating the images used to generate different images (looking the same, but different bytes), depending on the GD library. And it was difficult to force developers to run the same version of GD locally as on travis CI. For sure It would be better if we could test sparklines in UI tests... 

* the min/max dots are in my opinion quite useful to have as it helps scan the sparklines quickly and provides insights

* maybe we could just rename the `Object` class to another name to fix the issue, and keep our current sparklines with the min/max dots?

**Comment by sgiehl** [comment]
Those min/max dots should not be the only reason not to search for a new library. We should avoid to continue using such an old and not maintained library. 

I would agree with renaming the Object class to have a quick fix for PHP 7.2. But we should still consider replacing the old lib.

**Comment by Findus23** [comment]
Hi, I opened an feature request in the library, asking for an easy way to add dots:
https://github.com/davaxi/Sparkline/issues/2

**Comment by Findus23** [comment]
grafik

Now with minimum and maximum dots. I left out the end dot as I hope everyone is able to quickly see the end of the graph without it.

Hopefully my pull request (https://github.com/davaxi/Sparkline/pull/3) gets merged, so we can use the upstream version.

**Comment by sgiehl** [comment]
awesome! One thing that came to my mind when looking at the UI test results: Does it make sense to show the maximum and minimum dot when they both have the same value. If a sparkline always has 0 value (or any other constant value) it imho doesn't make that much sense to show the dots

**Comment by Findus23** [comment]
Now there are no min/max dots if they are the same.

In addition I added a padding to the top of the graph as otherwise the maximum would be always the top of the image and the dot was cut of

And I increased the image resolution, maybe it's looking better this way.

We could change the style a bit to better match the Piwik style, but I'm the wrong person to ask.

**Comment by mattab** [comment]
Nice @Findus23 that you added the feature upstream!

> And I increased the image resolution, maybe it's looking better this way.

Does it mean the sparkline will be "retina" compatible and look great on high res screen? that would be a nice benefit :+1:

**Comment by mattab** [comment]
I tried the PR and after running `composer install`, when loading a sparkline image, i get:

```
Invalid hexadecimal value

Stack trace

Invalid hexadecimal value

Go Back | Go to Piwik Enterprise
Stack trace

#0 vendor/davaxi/sparkline/src/Sparkline.php(271): Davaxi\Sparkline->colorHexToRGB(NULL)
#1 core/Visualization/Sparkline.php(143): Davaxi\Sparkline->setFillColorHex(NULL)
#2 core/Visualization/Sparkline.php(56): Piwik Enterprise\Visualization\Sparkline->setSparklineColors(Object(Davaxi\Sparkline))
#3 plugins/CoreVisualizations/Visualizations/Sparkline.php(59): Piwik Enterprise\Visualization\Sparkline->main()
#4 core/Plugin/Report.php(317): Piwik Enterprise\Plugins\CoreVisualizations\Visualizations\Sparkline->render()
#5 plugins/CoreHome/Controller.php(58): Piwik Enterprise\Plugin\Report->render()
#6 [internal function]: Piwik Enterprise\Plugins\CoreHome\Controller->renderReportWidget(Object(Piwik\Plugins\API\Reports\Get))
#7 core/FrontController.php(544): call_user_func_array(Array, Array)
#8 core/FrontController.php(137): Piwik Enterprise\FrontController->doDispatch('API', 'get', Array)
#9 core/dispatch.php(34): Piwik Enterprise\FrontController->dispatch()
#10 index.php(27): require_once('/home/matt/dev/...')
#11 {main}
```

**Comment by mattab** [comment]
Also got a Notification showing `WARNING: /home/matt/dev/piwik-master/core/Visualization/Sparkline.php(143): Notice - Undefined index: fillColor - Piwik 3.2.0 - Please report this message in the Piwik forums: https://forum.piwik.org (please do a search first as it might have been reported already)`

**Comment by mattab** [comment]
To keep things simple, we should try to make sparklines look exactly the same (sizes, colors, thickness of lines, size of dots, etc.). Currently they look quite different:

not same

**Comment by Findus23** [comment]
Seems like this pull request already fixes #12271, as the new library displays the following on empty data:
grafik

**Comment by Findus23** [comment]
I am not sure why the new color (fillcolor) won't get submitted for you.
I kept the (interesting 😄) way of defining the colors via CSS and just added a new element.
https://github.com/piwik/piwik/pull/12066/files#diff-cfb90e629f5b699cd4245bb8c5267632R10
Have you regenerated the CSS and Javascript files?

The difference in style in your screenshot is because of the doubling of the resolution as the size doesn't seem to be fixed there. I'll check that

**Comment by sgiehl** [comment]
@Findus23  I guess @mattab meant they should look exactly like before:
sparline

**Comment by Findus23** [comment]
As my pull request got merged and a new version got released, I changed the composer.json to point to the original library.

**Comment by sgiehl** [comment]
Would be awesome to merge that for one of the coming versions. 
Would you adjust the sparklines so they look the same as before? So line color back to `#162C4A`
Afterwards we can update the screenshots to be able to merge afterwards

**Comment by Findus23** [comment]
Fixed and update the UI tests

**Comment by mattab** [comment]
* Haven't yet tested in details but looking at the UI screenshots, the dots are cut off. Could we add some padding inside the image to prevent any dots being cut off?
* are the colors of the sparklines assigned from the URL parameters? as they could be themed and the colors are passed to the sparkline URL eg. `&colors={"backgroundColor"%3A"%23ffffff"%2C"lineColor"%3A"%23162c4a"%2C"minPointColor"%3A"%23ff7f7f"%2C"maxPointColor"%3A"%2375bf7c"%2C"lastPointColor"%3A"%2355aaff"}`

**Comment by mattab** [comment]
* also ideally we need a blue dot to show where the last value stands (ie. current selected period) so it's visually clear where the last value stands VS min/max

**Comment by Findus23** [comment]
- padding
Last time I tried, I failed to add a padding, so I only added it vertically. I’ll need to take a deeper look into the library to fix that. 
- styling
The graphs are coloured via the parameters which are created via the css styling. So any theme can change them by modifying the css (as before)
- lastpoint
I wasn’t sure if it was necessary so I haven’t added them yet. But that shouldn’t be that hard (apart from the padding issue)

**Comment by mattab** [comment]
Thanks for the update! 
lastPoint and adding some padding are I reckon needed for merge the PR and not lose features/styles :+1:

**Comment by Findus23** [comment]
@mattab, I think now I have all features.
I want to wait until my changes are merged upstream, but otherwise I think this is good to merge.

Personaly I'd prefer a more minimalistic style, so I think I'll publish a minimal theme looking similar to https://github.com/piwik/piwik/pull/12066/commits/43ee41be0cae9049cfffb70f23ec0ab8457e8a2a

That would also be a nice demo on how to style dynamic elements in Piwik.

**Comment by sgiehl** [comment]
Change https://github.com/piwik/piwik/pull/12066/commits/11073b8043228c6a045547459703324f453d76d2#diff-11ba2cc008adc397ffd66bf353d4a014 seems incorrect

**Comment by Findus23** [comment]
@sgiehl 
I am not sure what went wrong, but I fixed it now

**Comment by Findus23** [comment]
@sgiehl, @mattab 
davaxi has released a new version with all features needed by us and improved testing.

I have updated the pull request and apart from the UI test this should be ready to merge

grafik

I have also created a theme with an more minimalistic style (so the theme marketplace doesn't look that empty)
https://github.com/Findus23/AlternativeSparklines
grafik

**Comment by Findus23** [comment]
Ah I forgot that a `git fetch` isn't enough before reverting the conflicting files

**Comment by Findus23** [comment]
@mattab 

Is there anything that needs to be changed (apart from the UI tests) before this can be merged.

BTW: I noticed that this library was already included in Piwik 0.1.1, so it seems like I'm removing one of the oldest parts of Piwik 🙂

File: core/Visualization/Sparkline.php
Comment (sgiehl) [inline core/Visualization/Sparkline.php]: That peace of code is a bit outdated and won't work for all languages. Some languages use different characters to present decimals or percent values.
At first I thought about adding a `$this->requestConfig->request_parameters_to_modify['format_metrics'] = 0;` to the `render()` method of the Sparkline ViewDataTable, so the numbers aren't formatted at all, as that would remove any possible custom format, but that would result in decimal numbers instead of percent values (e.g. 0.53 instead of 53%), which then wouldn't show up correctly in the sparkline. Can't think of a good solution to handle the decimal percent values so the show up better.
Alternatively we could create a new method in NumberFormatter to convert a formatted number back to english format. Like https://github.com/matomo-org/matomo/blob/3.x-dev/core/NumberFormatter.php#L296-L306 does, but with flipped and reversed replacement values. But that would be more like a workaround.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left some comments. Besides that it looks good to me.
As the new library brings a lot of unneeded files with it (like tests and even the autoload.php, which isn't used with composer), we need to update our build script to remove them. Maybe you could already prepare a PR for that.

File: core/Visualization/Sparkline.php
Comment (Findus23) [reply core/Visualization/Sparkline.php]: To be honest, I haven't thought a lot about it, but kept it the same [as it was before](https://github.com/matomo-org/matomo/blob/3.x-dev/core/Visualization/Sparkline.php#L116)

Is it guaranteed that the decimal value will always have two decimal points? Then one could simply multiply it by 100.

File: tests/PHPUnit/Framework/TestingEnvironmentManipulator.php
Comment (Findus23) [reply tests/PHPUnit/Framework/TestingEnvironmentManipulator.php]: Done in https://github.com/matomo-org/matomo/pull/12066/commits/4da90daacf87e84770967b20d84c0ac0ee76793b

**Review by sgiehl — APPROVED** [review]
Rebased the branch and added a little improvements. The last UI results are looking good, so guess we could update them and merge...

**Pull request merged**
